src/object.d

```diff
@@ -4881,3 +4881,61 @@ void __ArrayDtor(T)(T[] a)
     foreach_reverse (ref T e; a)
         e.__xdtor();
 }
+
+/***********************************************************
+ * The compiler lowers expressions of `cast(TTo[])TFrom[]` to
+ * this implementation.
+ * Params:
+ *  from = the array to reinterpret-cast
+ * Returns:
+ *  `from` reinterpreted as `TTo[]`
+ */
+TTo[] __ArrayCast(TFrom, TTo)(TFrom[] from) @trusted
+{
+    const auto nbytes = from.length * TFrom.sizeof;
+
+    if ((nbytes % TTo.sizeof) != 0)
+    {
+        version(D_Exceptions)
+        {
+            import core.exception : onArrayMisalignmentError;
+            onArrayMisalignmentError();
```

Comment:
I think we should pass more informations to this function.

I'd like to see a message similar to this:
```
Error: Cannot cast `int[5]` to `ulong[]`, as size `20` is not aligned on `8` bytes.
```

It does not need to allocate. If you pass the elements separately, you can then provide a `toString` overload [which accepts a sink](https://github.com/dlang/druntime/blob/251cab752c25b9dcaf5bacb1344d2a26b49ed80c/src/object.d#L2635-L2642).

Let's not repeat the mistake that is array bounds mismatch's error message.

Reply:
Is your suggestion to have the `Error`'s `msg` say "Array cast misalignment", and then have the overloaded `toString` function return the more elaborate error message?

Reply:
I forgot people might be using `msg` directly... Although for `Error` it's unlikely. I think your suggestion makes sense.